core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -130,9 +130,9 @@ public void applyStartedShards(StartedRerouteAllocation allocation) {
     }
 
     public void applyFailedShards(FailedRerouteAllocation allocation) {
-        for (ShardRouting shard : allocation.failedShards()) {
-            Releasables.close(asyncFetchStarted.remove(shard.shardId()));
-            Releasables.close(asyncFetchStore.remove(shard.shardId()));
+        for (FailedRerouteAllocation.FailedShard shard : allocation.failedShards()) {
+            Releasables.close(asyncFetchStarted.remove(shard.getShard().shardId()));
+            Releasables.close(asyncFetchStore.remove(shard.getShard().shardId()));
```

Comment:
any chance that one of those throws an exception? if so should we catch and continue?